keps/sig-etcd/4578-server-feature-gate/README.md

```diff
@@ -0,0 +1,364 @@
+
+## Design Details
+
+### New APIs
+
+A new `--feature-gates` command line argument would be added to start the etcd server, with format like `--feature-gates=featureA=true,featureB=false`. The flag can also be set in the `config-file`.
+
+New Prometheus gauge metrics would be added to monitor and query if a feature is enabled for the server. 
+For easier usage in Kubernetes, we will also add a grpc backed `clientv3` function and a new `etcdctl` command.
+
+The metrics could look like:
+* `etcd_feature_enabled{name="featureName",stage="Alpha"} 1` if the feature is enabled.
+
+For grpc, new rpc type would be added to the `Maintenance` service.
+
+```proto
+service Maintenance {
+  ...
+  rpc FeatureGateStatus(FeatureGateStatusRequest) returns (FeatureGateStatusResponse) {
+    option (google.api.http) = {
+      post: "/v3/maintenance/featuregate"
+      body: "*"
+    };
+  }
+  ...
+}
+message Feature {
+  string name = 1;
+  bool enabled = 2;
+}
+message FeatureGateStatusRequest {
+  repeated string features = 2;
+}
+message FeatureGateStatusResponse {
+  ResponseHeader header = 1;
+  repeated Feature features = 3;
+}
+```
+
+The clientv3 function would look like:
+
+```go
+package clientv3
+
+type Maintenance interface {
+  FeatureGateStatus(ctx context.Context, features ...string) (*FeatureGateStatusResponse, error)
+}
+```
+
+The `etcdctl` commands could look like:
+* `etcdctl endpoint featuregate $featureName` returns true if the feature is enabled
+
+### Feature Gate
+
+We will use the new `k8s.io/component-base/featuregate.VersionedSpecs`(introduced in [kep-4330](https://github.com/kubernetes/enhancements/tree/master/keps/sig-architecture/4330-compatibility-versions)) to register and track the features along with their lifecycle at different release versions. 
+
+```go
+defaultServerFeatureGates := map[Feature]VersionedSpecs {
+		featureA: VersionedSpecs{
+			{Version: mustParseVersion("3.6"), Default: false, PreRelease: Beta},
+			{Version: mustParseVersion("3.7"), Default: true, PreRelease: GA},
+		},
+		featureB: VersionedSpecs{
+			{Version: mustParseVersion("3.7"), Default: false, PreRelease: Alpha},
+		},
+}
+```
+
+The feature gates for a server can only be set with the `--feature-gates`flag or in the `config-file` during startup. We do not support dynamically changing the feature gates when the server is running.
+
+The `ServerConfig` struct will have a new `featureGate k8s.io/component-base/featuregate.FeatureGate`(immutable) field. `EtcdServer` would have an interface of `FeatureEnabled(key Feature) bool`, and it can be referenced throughout the server.
+
+```go
+type ServerConfig struct {
+  ...
+  // ServerFeatureGate server level feature gate
+  ServerFeatureGate featuregate.FeatureGate
+  ...
+}
+
+func (s *EtcdServer) FeatureEnabled(key Feature) bool
+```
+
+### Feature Stages
+
+Following the convention of Kubernetes, a feature can go through a lifecycle of Alpha → Beta → GA → Deprecated.
```

Comment:
Is this new convention you are proposing? Can you reference any existing documentation and describe how it is expanded or adjusted.

Reply:
The feature stage `Properties` are copied from Kubernetes. I added the `Minimum Graduation Criteria`

Reply:
Please document them in KEP.